### PR TITLE
LUCENE-8904: enhance Nori DictionaryBuilder tool

### PR DESCRIPTION
**Pull request by danmuzi**
It is the Nori version of LUCENE-8863.
This patch has two changes.
1) Improve exception handling
2) Enable external dictionary for testing

Please check the following link:
https://issues.apache.org/jira/browse/LUCENE-8904

### CODE REVIEW

File: lucene/analysis/nori/src/tools/java/org/apache/lucene/analysis/ko/util/BinaryDictionaryWriter.java [hunk 1]
```diff
@@ -137,14 +139,17 @@ public int put(String[] entry) {
       flags |= BinaryDictionary.HAS_READING;
     }
 
-    assert leftId < 8192; // there are still unused bits
-    assert posType.ordinal() < 4;
+    if (leftId >= ID_LIMIT) {
+      throw new IllegalArgumentException("leftId >= " + ID_LIMIT + ": " + leftId);
+    }
+    if (posType.ordinal() >= 4) {
+      throw new IllegalArgumentException("posType.ordinal() >= " + 4 + ": " + posType.ordinal());
```
Comment (msokolov) [inline lucene/analysis/nori/src/tools/java/org/apache/lucene/analysis/ko/util/BinaryDictionaryWriter.java]: Maybe also log posType itself? If this ever happens, that should help

File: lucene/analysis/nori/src/tools/java/org/apache/lucene/analysis/ko/util/BinaryDictionaryWriter.java [hunk 2]
```diff
@@ -137,14 +139,17 @@ public int put(String[] entry) {
...
+    }
     buffer.putShort((short)(leftId << 2 | posType.ordinal()));
     buffer.putShort((short) (rightId << 2 | flags));
     buffer.putShort(wordCost);
 
     if (posType == POS.Type.MORPHEME) {
-      assert leftPOS == rightPOS;
```
Comment (msokolov) [inline lucene/analysis/nori/src/tools/java/org/apache/lucene/analysis/ko/util/BinaryDictionaryWriter.java]: I don't understand how this works in Korean - are you sure it's OK to remove?

File: lucene/analysis/nori/src/tools/java/org/apache/lucene/analysis/ko/util/BinaryDictionaryWriter.java (hunk 1 above)
Comment (danmuzi) [reply lucene/analysis/nori/src/tools/java/org/apache/lucene/analysis/ko/util/BinaryDictionaryWriter.java]: +1.

Which log style is better?
1) `throw new IllegalArgumentException("posType.ordinal() >= " + 4 + ": " + posType.name());`
2) `throw new IllegalArgumentException("posType should be MORPHOME or COMPOUND or INFLECT or PREANALYSIS" + ": " + posType.name());`

File: lucene/analysis/nori/src/tools/java/org/apache/lucene/analysis/ko/util/BinaryDictionaryWriter.java (hunk 2 above)
Comment (danmuzi) [reply lucene/analysis/nori/src/tools/java/org/apache/lucene/analysis/ko/util/BinaryDictionaryWriter.java]: It is an unnecessary condition.
The leftPOS and rightPOS are determined on lines 83-90.
If posType is POS.Type.MORPHEME, leftPOS and rightPOS are always the same.

**Comment by danmuzi** [comment]
Thank you for your review! @msokolov
I left some replies :D

File: lucene/analysis/nori/src/tools/java/org/apache/lucene/analysis/ko/util/BinaryDictionaryWriter.java (hunk 2 above)
Comment (msokolov) [reply lucene/analysis/nori/src/tools/java/org/apache/lucene/analysis/ko/util/BinaryDictionaryWriter.java]: Well, assertions are always unnecessary! I think their value is exactly in helping ensure that the correct code path was followed to get here, and confirming what we know to be true by analysis or anyway knowledge, in fact holds in practice (empirically). For example, if I were to somehow mess up the code above that guarantees this, due to my ignorance, this assertion would helpfully let me know. So I think we should keep it?

File: lucene/analysis/nori/src/tools/java/org/apache/lucene/analysis/ko/util/BinaryDictionaryWriter.java (hunk 1 above)
Comment (msokolov) [reply lucene/analysis/nori/src/tools/java/org/apache/lucene/analysis/ko/util/BinaryDictionaryWriter.java]: (1), please!

File: lucene/analysis/nori/src/tools/java/org/apache/lucene/analysis/ko/util/BinaryDictionaryWriter.java (hunk 2 above)
Comment (danmuzi) [reply lucene/analysis/nori/src/tools/java/org/apache/lucene/analysis/ko/util/BinaryDictionaryWriter.java]: OK :)
I understand what you mean and I'll keep it.

File: lucene/analysis/nori/src/tools/java/org/apache/lucene/analysis/ko/util/BinaryDictionaryWriter.java (hunk 1 above)
Comment (danmuzi) [reply lucene/analysis/nori/src/tools/java/org/apache/lucene/analysis/ko/util/BinaryDictionaryWriter.java]: Thanks :)

**Comment by danmuzi** [comment]
Thank you for your review! @msokolov 
I'll merge(Squash and merge) this PR and close.

**Pull request merged**